File: apps/web/lib/types/coding.ts
// Coding challenge type definitions

export interface Example {
  input: string;
  output: string;
  explanation?: string;
}

export interface CodingProblem {
  id: string;
  title: string;
  difficulty: "easy" | "medium" | "hard";
  description: string;
  examples: Example[];
  constraints: string[];
  starterCode: Record<string, string>;
  tags: string[];
}

export interface CodeSubmission {
  problemId: string;
  code: string;
  language: string;
}

export interface CodeEvaluationResult {
  correct: boolean;
  score: number;
  feedback: string;
  analysis?: {
    correctness: number;
    edgeCaseHandling: number;
    codeQuality: number;
    complexity: number;
  };
}

export type SupportedLanguage = "python" | "javascript" | "typescript" | "java" | "cpp" | "go";

export const LANGUAGE_OPTIONS: { value: SupportedLanguage; label: string; monacoId: string }[] = [
  { value: "python", label: "Python", monacoId: "python" },
  { value: "javascript", label: "JavaScript", monacoId: "javascript" },
  { value: "typescript", label: "TypeScript", monacoId: "typescript" },
  { value: "java", label: "Java", monacoId: "java" },
  { value: "cpp", label: "C++", monacoId: "cpp" },
  { value: "go", label: "Go", monacoId: "go" },
];

export const DEFAULT_LANGUAGE: SupportedLanguage = "python";

File: services/api/app/schemas/coding.py
"""Schemas for coding challenge feature."""

from typing import Literal

from pydantic import BaseModel, Field


class Example(BaseModel):
    """Example test case for a coding problem."""

    input: str
    output: str
    explanation: str | None = None


class CodingProblem(BaseModel):
    """A coding challenge problem."""

    id: str
    title: str
    difficulty: Literal["easy", "medium", "hard"]
    description: str
    examples: list[Example]
    constraints: list[str]
    starter_code: dict[str, str]  # language -> starter code
    tags: list[str]


class CodeSubmission(BaseModel):
    """Code submission from the candidate."""

    problem_id: str
    code: str
    language: str


class CodeEvaluationAnalysis(BaseModel):
    """Detailed analysis of code evaluation."""

    correctness: int = Field(ge=0, le=100)
    edge_case_handling: int = Field(ge=0, le=100)
    code_quality: int = Field(ge=0, le=100)
    complexity: int = Field(ge=0, le=100)


class CodeEvaluationResult(BaseModel):
    """Result of evaluating submitted code."""

    correct: bool
    score: float
    feedback: str
    analysis: CodeEvaluationAnalysis | None = None
